### PR TITLE
tests: add tests for fsck; cmd/s-b/initramfs-mounts: fsck ubuntu-seed too

### PR DESCRIPTION
**Pull request by zyga**
File system checker is an essential part of reliability of ubuntu-core
systems. The pair of new tests exercise fsck.vfat in particular and how
it is automatically invoked by the boot base, in a manner that is not
explored, automatically, in case a file system was not unmounted
cleanly.

Incidentally the test uncovered a real failure on core20, whatever
part is responsible for verifying vfat during the boot process, before
that partition is mounted, is not executing the same way it does on
core16 and core18.

Signed-off-by: Zygmunt Krynicki <me@zygoon.pl>

### CODE REVIEW

File: tests/core/fsck-vfat/task.yaml
Comment (zyga) [inline tests/core/fsck-vfat/task.yaml:47]: The dirty bit does not go away after unmount if it was initially dirty. We can extract the dirty bit with:

```
dd if=fat.img of=dirty skip=37 bs=1 count=1 conv=notrunc
```
We can use this in fsck-on-boot test, to unmount the relevant partition and extract the dirty bit. This test is superior to just scanning the kernel logs for a message that may or may not be there over time, and relies on early-boot logging handover implemented by systemd

File: tests/core/fsck-on-boot/task.yaml
Comment (zyga) [inline tests/core/fsck-on-boot/task.yaml:69]: We should implement the trick mentioned below, I suspect this is not working anywhere, and that only core20, thanks to systemd inside initrd, we just get the log across.

**Review by cmatsuoka — APPROVED** [review]
Thanks for this test. Added a question about the offset on FAT32 which is inconsistent between comment and implementation.

File: tests/core/fsck-on-boot/task.yaml
Comment (bboozzoo) [inline tests/core/fsck-on-boot/task.yaml:40]: I've written a small tool back when we were investigating the corruption uboot.env when we got 2 files, one with uppercase and the other with lowercase names. The tool is here: https://gist.github.com/bboozzoo/e68254507eef4673dd8c6f9b82f65d92 would it make sense to use it as a starting point to try and corrupt the fs in a clever way, that would make it non-mountable?

File: tests/core/fsck-on-boot/task.yaml
Comment (bboozzoo) [inline tests/core/fsck-on-boot/task.yaml:69]: I'd rather we try to corrupt the system in a way that renders it non-mountable, but still can be fixed with fsck and mounter afterwards.

File: tests/core/fsck-on-boot/task.yaml
Comment (zyga) [reply tests/core/fsck-on-boot/task.yaml:40]: That's an interesting test but I wanted to prove something easier, that a dirty flag can be cleared because there is _some_ tool in the boot chain that handles that. This test is sufficient for that and doesn't require any more advanced corruption. I would love to add more comprehensive test but my point is this test is useful and (hopefully) precise in intent.

File: tests/core/fsck-on-boot/task.yaml
Comment (zyga) [reply tests/core/fsck-on-boot/task.yaml:69]: I believe that is a more serious test and I would like to keep it separate.

File: tests/core/fsck-on-boot/task.yaml
Comment (zyga) [reply tests/core/fsck-on-boot/task.yaml:31]: Yes, I think I was looking at amd64 above and wrote this wrongly somehow. The correct offset is 64.

File: tests/core/fsck-on-boot/task.yaml
Comment (zyga) [reply tests/core/fsck-on-boot/task.yaml:3]: I'm not sure what Ian worked on. I'll ask him next week.

**Comment by zyga** [comment]
The extra check I was thinking about shows that fsck is not invoked on boot on any core system. This is troubling.

**Comment by zyga** [comment]
After improving the test, it does pass on core18 and fails on core16 and core20. I suspect core16 is too old to fsck with systemd and core20 doesn't get a chance to rely on that feature, because snapd bootstrap logic has already mounted that partition.

File: tests/core/fsck-on-boot/task.yaml
Comment (sergiocazzolato) [inline tests/core/fsck-on-boot/task.yaml:15]: perhaps here you could use the check declared in systems.sh (is_core16_system)

File: tests/core/fsck-on-boot/task.yaml
Comment (sergiocazzolato) [inline tests/core/fsck-on-boot/task.yaml:86]: this is not needed, it is done automatically

File: tests/core/fsck-vfat/task.yaml
Comment (sergiocazzolato) [inline tests/core/fsck-vfat/task.yaml:68]: this is not needed

**Review by sergiocazzolato — APPROVED** [review]
Thanks for the tests, I left few comments (not blocking)

**Review by mvo5 — APPROVED** [review]
This looks good. Thanks for adding this extra line of defense

File: tests/core/fsck-on-boot/task.yaml
Comment (anonymouse64) [inline tests/core/fsck-on-boot/task.yaml:7]:
```suggestion
  the boot process is capable of detecting unclean vfat and fixing it before
```

File: tests/core/fsck-on-boot/task.yaml
Comment (anonymouse64) [inline tests/core/fsck-on-boot/task.yaml:9]: thanks for the details in the spread test! very nice

File: tests/core/fsck-on-boot/task.yaml
Comment (anonymouse64) [inline tests/core/fsck-on-boot/task.yaml:29]: we have other TODO's about doing this, the issue is that some operations require ubuntu-seed to be writable, but I think the thinking there was that we would just re-mount as writable to do the things we need, then re-mount as read-only

File: tests/core/fsck-on-boot/task.yaml
Comment (anonymouse64) [inline tests/core/fsck-on-boot/task.yaml:2]: lol the reason this doesn't work on uc20 is by design I had misunderstood the requirements of fsck for vfats, see

https://github.com/snapcore/snapd/blob/c8573086fe4ad024bf2b573527f64ec1a657487a/cmd/snap-bootstrap/cmd_initramfs_mounts.go#L478-L479

I will file a patch after this lands to change that and adjust this test accordingly, thanks for the test

File: tests/core/fsck-on-boot/task.yaml
Comment (anonymouse64) [inline tests/core/fsck-on-boot/task.yaml:72]:
```suggestion
      # Unmount vfat again and read the dirty flag manually. The kernel doesn't
```

File: tests/core/fsck-vfat/task.yaml
Comment (anonymouse64) [inline tests/core/fsck-vfat/task.yaml:13]: we should also add tests for `mkfs.vfat`, etc. too I think we had a bug where core20 didn't actually have `mkfs.vfat` that was only discovered when enabling the pi and would have been nice to catch earlier in the dev process (at that point we were running spread tests for amd64)

File: tests/core/fsck-vfat/task.yaml
Comment (anonymouse64) [inline tests/core/fsck-vfat/task.yaml:59]: `NOMATCH` here ?

**Review by anonymouse64 — APPROVED** [review]
lgtm, sorry for misunderstanding how fsck works, I can fix this for uc20 after this lands

File: tests/core/fsck-vfat/task.yaml
Comment (zyga) [reply tests/core/fsck-vfat/task.yaml:59]: Oh I forgot about this part. I'll look into this.

Comment (zyga) [reply tests/core/fsck-vfat/task.yaml:59]: Done, also adjusted the logic to make NOMATCH available elsewhere.

File: tests/core/fsck-on-boot/task.yaml
Comment (zyga) [reply tests/core/fsck-on-boot/task.yaml:29]: Yeah, I think this is preferred. Ideally we'd mount the partition as read only to begin with. Then re-mount as writable only inside snapd, around specific operations. This should decrease the chance of corruption due to accidental power-cycle significantly.

**Comment by zyga** [comment]
The failure on core16 is surprising. There is an active fsck service there:
```
google:ubuntu-core-16-64 /etc/systemd/system# systemctl status systemd-fsck@dev-sda2.service
● systemd-fsck@dev-sda2.service - File System Check on /dev/sda2
   Loaded: loaded (/lib/systemd/system/systemd-fsck@.service; static; vendor preset: enabled)
   Active: active (exited) since Mon 2020-10-19 12:32:53 UTC; 41min ago
     Docs: man:systemd-fsck@.service(8)
 Main PID: 1328 (code=exited, status=0/SUCCESS)

Oct 19 12:32:53 localhost.localdomain systemd[1]: Starting File System Check on /dev/sda2...
Oct 19 12:32:53 localhost.localdomain systemd-fsck[1328]: fsck.fat 3.0.28 (2015-05-16)
Oct 19 12:32:53 localhost.localdomain systemd-fsck[1328]: /dev/sda2: 8 files, 4840/100792 clusters
Oct 19 12:32:53 localhost.localdomain systemd[1]: Started File System Check on /dev/sda2.
```

**Comment by zyga** [comment]
Core16 passes, the test was incorrect. Now only core20 fails.

File: tests/core/fsck-on-boot/task.yaml
Comment (anonymouse64) [reply tests/core/fsck-on-boot/task.yaml]: I pushed the patch here, this should be good

**Review by anonymouse64 — APPROVED** [review]
still lgtm (though someone else should +1 my uc20 changes I just pushed)

**Comment by zyga** [comment]
I'll do a pass to adjust what Sergio has commented on and will merge when green.

**Review by mvo5 — APPROVED** [review]

File: tests/core/fsck-on-boot/task.yaml
Comment (zyga) [reply tests/core/fsck-on-boot/task.yaml:13]: I used your new os.query tool instead.

**Comment by zyga** [comment]
I'll add mkfs tests in a separate PR, LGTM.

File: tests/core/fsck-vfat/task.yaml
Comment (zyga) [reply tests/core/fsck-vfat/task.yaml:13]: I'll do this in a follow-up

File: tests/core/fsck-on-boot/task.yaml
Comment (bboozzoo) [inline tests/core/fsck-on-boot/task.yaml:12]: looks like an unresolved conflict here?

Comment (zyga) [reply tests/core/fsck-on-boot/task.yaml]: Sorry about that, my mistake.

**Review by bboozzoo — APPROVED** [review]
Achievement unlocked, 5 +1s.

**Pull request merged**
